Micronaut v4 breaks when using `suspend` parameter lambda in test

### Expected Behavior

If a test passes without `@MicronautTest` present, it should pass with `@MicronautTest` present. If a test uses a custom Kotlin receiver context, it shouldn't fail merely because Micronaut processed it.

### Actual Behaviour

The attached repro fails, but only if `@MicronautTest` is present on the test. The test is not exotic, it just uses a parent method ("`test`") which is called from the concrete test to establish a receiver context.

### Steps To Reproduce

1. Generate Micronaut sample with Kotlin and KSP added (`4.0.0-M4`)
2. Add dependencies for KotlinX coroutines, KotlinX coroutines test
3. Add a base class and interface which create a parent test method that, when called from a test, establishes a custom Kotlin receiver context (sounds complex, but it's not, see below for sample)
4. Observe that tests fail when extending the base class with a `@MicronautTest` annotation, but not without

What I mean by "establishes a custom receiver context":

```kotlin
import kotlinx.coroutines.test.runTest

/** The "test context" describes the layout of the receiver for a test */
interface TestContext {
  /** Utility provided to tests. */
  fun someUtility()
}

/** The "test case" base interface defines the entrypoint for tests which create the context */
interface TestCase<Context> where Context : TestContext {
  /** Entrypoint for a suspending `Context`-enabled test. */
  fun test(fn: suspend Context.() → Unit)
}

/** Abstract base that concrete tests can extend, to simply call `test` */
abstract class BaseTestCase : TestCase<TestContext> {
  override fun test(fn: TestContext.() → Unit) = runTest {
    fn.invoke(object : TestContext {
      override fun someUtility() { /* ... */ }
    })
  }
}

/** Concrete test example: working */
class WorkingTest: BaseTestCase() {
  @Test fun someTest() = test {
    // ... assertions (in here i can call `someUtility` because we're in the context of `TestContext`)
  }
}

/** Concrete test example: failing */
@MicronautTest class FailingTest: BaseTestCase() {
  @Test fun someTest() = test {
    // ... assertions (in here i can call `someUtility` because we're in the context of `TestContext`)
  }
}
```

The reproducer project makes this clearer with a few working test cases, and one failing case. To see this in action, run:
```
./gradlew test
```

Observe failure:

```
Exception java.lang.NoClassDefFoundError: kotlin/coroutines/SuspendFunction1
```

<details>
<summary>Expand for stacktrace</summary>
<pre>
    io.micronaut.context.exceptions.BeanInstantiationException: Bean definition [com.example.FailingTest] could not be loaded: kotlin/coroutines/SuspendFunction1
        at app//io.micronaut.context.DefaultBeanContext$BeanDefinitionProducer.getDefinition(DefaultBeanContext.java:4158)
        at app//io.micronaut.context.DefaultBeanContext.collectBeanCandidates(DefaultBeanContext.java:2138)
        at app//io.micronaut.context.DefaultBeanContext.findBeanCandidates(DefaultBeanContext.java:2113)
        at app//io.micronaut.context.DefaultBeanContext.findConcreteCandidateNoCache(DefaultBeanContext.java:3113)
        at app//io.micronaut.context.DefaultBeanContext.findConcreteCandidate(DefaultBeanContext.java:3097)
        at app//io.micronaut.context.DefaultBeanContext.findBeanDefinition(DefaultBeanContext.java:761)
        at app//io.micronaut.context.DefaultBeanContext.findBeanDefinition(DefaultBeanContext.java:770)
        at app//io.micronaut.context.BeanDefinitionRegistry.findBeanDefinition(BeanDefinitionRegistry.java:595)
        at app//io.micronaut.test.extensions.AbstractMicronautExtension.beforeClass(AbstractMicronautExtension.java:328)
        at app//io.micronaut.test.extensions.junit5.MicronautJunit5Extension.beforeAll(MicronautJunit5Extension.java:84)
        at app//org.junit.jupiter.engine.descriptor.ClassBasedTestDescriptor.lambda$invokeBeforeAllCallbacks$12(ClassBasedTestDescriptor.java:395)
        at app//org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
        at app//org.junit.jupiter.engine.descriptor.ClassBasedTestDescriptor.invokeBeforeAllCallbacks(ClassBasedTestDescriptor.java:395)
        at app//org.junit.jupiter.engine.descriptor.ClassBasedTestDescriptor.before(ClassBasedTestDescriptor.java:211)
        at app//org.junit.jupiter.engine.descriptor.ClassBasedTestDescriptor.before(ClassBasedTestDescriptor.java:84)
        at app//org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$6(NodeTestTask.java:148)
        at app//org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
        at app//org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:141)
        at app//org.junit.platform.engine.support.hierarchical.Node.around(Node.java:137)
        at app//org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$9(NodeTestTask.java:139)
        at app//org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
        at app//org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:138)
        at app//org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:95)
        at java.base@17.0.7/java.util.ArrayList.forEach(ArrayList.java:1511)
        at app//org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.invokeAll(SameThreadHierarchicalTestExecutorService.java:41)
        at app//org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$6(NodeTestTask.java:155)
        at app//org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
        at app//org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:141)
        at app//org.junit.platform.engine.support.hierarchical.Node.around(Node.java:137)
        at app//org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$9(NodeTestTask.java:139)
        at app//org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
        at app//org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:138)
        at app//org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:95)
        at app//org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.submit(SameThreadHierarchicalTestExecutorService.java:35)
        at app//org.junit.platform.engine.support.hierarchical.HierarchicalTestExecutor.execute(HierarchicalTestExecutor.java:57)
        at app//org.junit.platform.engine.support.hierarchical.HierarchicalTestEngine.execute(HierarchicalTestEngine.java:54)
        at app//org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:147)
        at app//org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:127)
        at app//org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:90)
        at app//org.junit.platform.launcher.core.EngineExecutionOrchestrator.lambda$execute$0(EngineExecutionOrchestrator.java:55)
        at app//org.junit.platform.launcher.core.EngineExecutionOrchestrator.withInterceptedStreams(EngineExecutionOrchestrator.java:102)
        at app//org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:54)
        at app//org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:114)
        at app//org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:86)
        at app//org.junit.platform.launcher.core.DefaultLauncherSession$DelegatingLauncher.execute(DefaultLauncherSession.java:86)
        at org.gradle.api.internal.tasks.testing.junitplatform.JUnitPlatformTestClassProcessor$CollectAllTestClassesExecutor.processAllTestClasses(JUnitPlatformTestClassProcessor.java:110)
        at org.gradle.api.internal.tasks.testing.junitplatform.JUnitPlatformTestClassProcessor$CollectAllTestClassesExecutor.access$000(JUnitPlatformTestClassProcessor.java:90)
        at org.gradle.api.internal.tasks.testing.junitplatform.JUnitPlatformTestClassProcessor.stop(JUnitPlatformTestClassProcessor.java:85)
        at org.gradle.api.internal.tasks.testing.SuiteTestClassProcessor.stop(SuiteTestClassProcessor.java:62)
        at java.base@17.0.7/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at java.base@17.0.7/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
        at java.base@17.0.7/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.base@17.0.7/java.lang.reflect.Method.invoke(Method.java:568)
        at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:36)
        at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:24)
        at org.gradle.internal.dispatch.ContextClassLoaderDispatch.dispatch(ContextClassLoaderDispatch.java:33)
        at org.gradle.internal.dispatch.ProxyDispatchAdapter$DispatchingInvocationHandler.invoke(ProxyDispatchAdapter.java:94)
        at jdk.proxy1/jdk.proxy1.$Proxy2.stop(Unknown Source)
        at org.gradle.api.internal.tasks.testing.worker.TestWorker$3.run(TestWorker.java:193)
        at org.gradle.api.internal.tasks.testing.worker.TestWorker.executeAndMaintainThreadName(TestWorker.java:129)
        at org.gradle.api.internal.tasks.testing.worker.TestWorker.execute(TestWorker.java:100)
        at org.gradle.api.internal.tasks.testing.worker.TestWorker.execute(TestWorker.java:60)
        at org.gradle.process.internal.worker.child.ActionExecutionWorker.execute(ActionExecutionWorker.java:56)
        at org.gradle.process.internal.worker.child.SystemApplicationClassLoaderWorker.call(SystemApplicationClassLoaderWorker.java:113)
        at org.gradle.process.internal.worker.child.SystemApplicationClassLoaderWorker.call(SystemApplicationClassLoaderWorker.java:65)
        at app//worker.org.gradle.process.internal.worker.GradleWorkerMain.run(GradleWorkerMain.java:69)
        at app//worker.org.gradle.process.internal.worker.GradleWorkerMain.main(GradleWorkerMain.java:74)
    Caused by: java.lang.NoClassDefFoundError: kotlin/coroutines/SuspendFunction1
        at com.example.$FailingTest$Definition$Exec.$metadata$test(Unknown Source)
        at com.example.$FailingTest$Definition$Exec.<clinit>(Unknown Source)
        at com.example.$FailingTest$Definition.<init>(Unknown Source)
        at com.example.$FailingTest$Definition.<init>(Unknown Source)
        at com.example.$FailingTest$Definition$Reference.load(Unknown Source)
        at io.micronaut.context.AbstractInitializableBeanDefinitionReference.load(AbstractInitializableBeanDefinitionReference.java:178)
        at io.micronaut.context.DefaultBeanContext$BeanDefinitionProducer.getDefinition(DefaultBeanContext.java:4153)
        ... 66 more
    Caused by: java.lang.ClassNotFoundException: kotlin.coroutines.SuspendFunction1
        at java.base/jdk.internal.loader.BuiltinClassLoader.loadClass(BuiltinClassLoader.java:641)
        at java.base/jdk.internal.loader.ClassLoaders$AppClassLoader.loadClass(ClassLoaders.java:188)
        at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:520)
        ... 73 more
</pre>
</details>

### Environment Information

- Operating System: macOS Ventura `13.4.1`
- JVM: Zulu 19 (see below)

```
openjdk 19.0.2 2023-01-17
OpenJDK Runtime Environment Zulu19.32+13-CA (build 19.0.2+7)
OpenJDK 64-Bit Server VM Zulu19.32+13-CA (build 19.0.2+7, mixed mode, sharing)
```

### Example Application

https://github.com/sgammon/micronaut-kotlin-coroutines-repro

### Version

4.0.0-M4

--- a/inject-kotlin/src/main/kotlin/io/micronaut/kotlin/processing/extensions.kt
+++ b/inject-kotlin/src/main/kotlin/io/micronaut/kotlin/processing/extensions.kt
@@ -20,6 +20,7 @@ import com.google.devtools.ksp.getJavaClassByName
 import com.google.devtools.ksp.processing.Resolver
 import com.google.devtools.ksp.symbol.*
 import io.micronaut.kotlin.processing.visitor.KotlinVisitorContext
+import org.objectweb.asm.Type
 
 @OptIn(KspExperimental::class)
 internal fun KSDeclaration.getBinaryName(resolver: Resolver, visitorContext: KotlinVisitorContext): String {
@@ -31,17 +32,24 @@ internal fun KSDeclaration.getBinaryName(resolver: Resolver, visitorContext: Kot
         }
     }
     val binaryName = resolver.mapKotlinNameToJava(declaration.qualifiedName!!)?.asString()
-    return if (binaryName != null) {
-        binaryName
-    } else {
-        val classDeclaration = declaration.getClassDeclaration(visitorContext)
-        val qn = classDeclaration.qualifiedName
-        if (qn != null) {
-            resolver.mapKotlinNameToJava(qn)?.asString() ?: computeName(declaration)
-        } else {
-            computeName(declaration)
+    if (binaryName != null) {
+        return binaryName
+    }
+    val classDeclaration = declaration.getClassDeclaration(visitorContext)
+    val qn = classDeclaration.qualifiedName
+    if (qn != null) {
+        val asString = resolver.mapKotlinNameToJava(qn)?.asString()
+        if (asString != null) {
+            return asString;
         }
     }
+    if (declaration is KSClassDeclaration) {
+        val signature = resolver.mapToJvmSignature(declaration)
+        if (signature != null) {
+            return Type.getType(signature).className
+        }
+    }
+    return computeName(declaration)
 }
 
 private fun computeName(declaration: KSDeclaration): String {

--- a/inject-kotlin/src/main/kotlin/io/micronaut/kotlin/processing/visitor/KotlinClassElement.kt
+++ b/inject-kotlin/src/main/kotlin/io/micronaut/kotlin/processing/visitor/KotlinClassElement.kt
@@ -139,9 +139,7 @@ internal open class KotlinClassElement(
         javaName?.asString() ?: declaration.qualifiedName!!.asString()
     }
 
-    private val internalName: String by lazy {
-        declaration.getBinaryName(visitorContext.resolver, visitorContext)
-    }
+    private val internalName = declaration.getBinaryName(visitorContext.resolver, visitorContext)
 
     private val resolvedInterfaces: Collection<ClassElement> by lazy {
         declaration.superTypes.map { it.resolve() }
@@ -509,14 +507,6 @@ internal open class KotlinClassElement(
             return type.kotlinType.starProjection().makeNullable().isAssignableFrom(kotlinType.starProjection().makeNullable())
         }
         return super.isAssignable(type)
-    }
-
-    override fun isPrimitive(): Boolean {
-        return isVoid
-    }
-
-    override fun isVoid(): Boolean {
-        return internalName == "kotlin.Unit"
     }
 
     override fun copyThis() = KotlinClassElement(

--- a/inject-kotlin/src/test/groovy/io/micronaut/kotlin/processing/beans/BeanDefinitionSpec.groovy
+++ b/inject-kotlin/src/test/groovy/io/micronaut/kotlin/processing/beans/BeanDefinitionSpec.groovy
@@ -24,6 +24,54 @@ import static io.micronaut.annotation.processing.test.KotlinCompiler.getBeanDefi
 
 class BeanDefinitionSpec extends Specification {
 
+    void "test bean suspend fn"() {
+        when:
+        def context = buildContext('''
+package test
+
+import io.micronaut.context.annotation.Executable
+import jakarta.inject.Singleton
+import kotlinx.coroutines.test.runTest
+import io.micronaut.runtime.EmbeddedApplication
+import io.micronaut.test.extensions.junit5.annotation.MicronautTest
+import jakarta.inject.Inject
+import kotlinx.coroutines.test.runTest
+import org.junit.jupiter.api.Assertions
+
+interface TestCase<Context> where Context : TestContext {
+    fun test(block: suspend Context.() -> Unit)
+}
+
+abstract class BaseTest : TestCase<TestContext> {
+    override fun test(block: suspend TestContext.() -> Unit) = runTest {
+        block.invoke(object : TestContext { /* nothing here */ })
+    }
+}
+
+@Singleton
+@Executable
+class FailingTest : BaseTest() {
+}
+
+interface TestContext
+
+''')
+
+        then:
+        context != null
+
+        when:
+        getBean(context, 'test.FailingTest')
+        def bd = getBeanDefinition(context, 'test.FailingTest')
+        def testMethod = bd.findPossibleMethods("test").findFirst().orElseThrow()
+        def param1 = testMethod.getArguments()[0]
+        then:
+        // essentially this is not correct as KSP will provide some kotlin.coroutines.SuspendFunction1 which is internally remapped to kotlin.jvm.functions.Function2
+        param1.type.name == "kotlin.jvm.functions.Function2"
+        param1.typeParameters[0].type.name == "test.TestContext"
+        param1.typeParameters[1].type.name == "kotlin.Unit"
+    }
+
     void "test bean annotated with @MicronautTest"() {
         when:
         def context = buildContext('''

--- a/inject-kotlin/src/test/groovy/io/micronaut/kotlin/processing/inject/ast/ClassElementSpec.groovy
+++ b/inject-kotlin/src/test/groovy/io/micronaut/kotlin/processing/inject/ast/ClassElementSpec.groovy
@@ -1723,7 +1723,6 @@ class MyBeanX {
             def element = ce.findMethod("save").get().getParameters()[0]
             element.getGenericType().simpleName == "MyBeanX"
             element.getType().simpleName == "Object"
-            element.getGenericType().isAssignable("test.MyBeanX")
         when:
             def genRepo = ce.getTypeArguments("test.GenericRepository")
         then:
@@ -1731,6 +1730,34 @@ class MyBeanX {
             genRepo.get("E").getSyntheticBeanProperties().size() == 1
             genRepo.get("E").getMethods().size() == 0
             genRepo.get("E").getFields().get(0).name == "name"
+    }
+
+    void "test interface placeholder 2 isAssignable"() {
+        when:
+        boolean isAssignable = buildClassElementMapped('test.MyRepoX', '''
+package test
+import io.micronaut.context.annotation.Prototype
+import java.util.List
+
+interface MyRepoX : RepoX<MyBeanX, Long>
+interface RepoX<E, ID> : GenericRepository<E, ID> {
+    fun <S : E> save(ent: S)
+}
+interface GenericRepository<E, ID>
+@Prototype
+class MyBeanX {
+    var name: String? = null
+}
+
+''', { ce ->
+            def element = ce.findMethod("save").get().getParameters()[0]
+            element.getGenericType().simpleName == "MyBeanX"
+            element.getType().simpleName == "Object"
+            element.getGenericType().isAssignable("test.MyBeanX")
+        })
+
+        then:
+            isAssignable
     }
 
     void "test abstract placeholder"() {

--- a/inject-kotlin/src/test/groovy/io/micronaut/kotlin/processing/visitor/KotlinReconstructionSpec.groovy
+++ b/inject-kotlin/src/test/groovy/io/micronaut/kotlin/processing/visitor/KotlinReconstructionSpec.groovy
@@ -50,7 +50,7 @@ class Lst<in E> {
     @Unroll("type var is #decl")
     def 'type vars declared on method'() {
         given:
-            def method = buildClassElementMapped("test.Test", """
+            def methodDecl = buildClassElementMapped("test.Test", """
 package test;
 
 import java.util.*;
@@ -60,10 +60,13 @@ abstract class Test<A> {
 }
 class Lst<in E> {
 }
-""", element -> element.<MethodElement> getEnclosedElement(ElementQuery.ALL_METHODS.named(s -> s == 'method')).get())
+""", {
+                element ->
+                    def method = element.<MethodElement> getEnclosedElement(ElementQuery.ALL_METHODS.named(s -> s == 'method')).get()
+                    return reconstructTypeSignature(method.declaredTypeVariables[0], true)
+            })
         expect:
-            reconstructTypeSignature(method.declaredTypeVariables[0], true) == decl
-
+            methodDecl == decl
         where:
             decl << [
 //                    'T',
@@ -137,7 +140,7 @@ abstract class Test<T> : $superType {
     @Unroll("type var is #decl")
     def 'type vars declared on type'() {
         given:
-            def element = buildClassElement("test.Test", """
+            def elementDecl = buildClassElementMapped("test.Test", """
 package test;
 
 import java.util.*;
@@ -147,10 +150,12 @@ abstract class Test<A, $decl> {
 
 class Lst<in E> {
 }
-""")
+""", element -> {
+                reconstructTypeSignature(element.declaredGenericPlaceholders[1], true)
+            })
 
         expect:
-            reconstructTypeSignature(element.declaredGenericPlaceholders[1], true) == decl
+            elementDecl == decl
 
         where:
             decl << [
@@ -170,7 +175,7 @@ class Lst<in E> {
     @Unroll('declaration is #decl')
     def 'fold type variable to null'() {
         given:
-            def classElement = buildClassElement("test.Test", """
+            def classDecl = buildClassElementMapped("test.Test", """
 package test;
 
 import java.util.*;
@@ -181,17 +186,19 @@ class Test<T> {
 
 class Lst<in E> {
 }
-""")
-            def fieldType = classElement.fields[0].type
+""", {
+                def fieldType = it.fields[0].type
+                return reconstructTypeSignature(fieldType.foldBoundGenericTypes {
+                    if (it != null && it.isGenericPlaceholder() && ((GenericPlaceholderElement) it).variableName == 'T') {
+                        return null
+                    } else {
+                        return it
+                    }
+                })
+            })
 
         expect:
-            reconstructTypeSignature(fieldType.foldBoundGenericTypes {
-                if (it != null && it.isGenericPlaceholder() && ((GenericPlaceholderElement) it).variableName == 'T') {
-                    return null
-                } else {
-                    return it
-                }
-            }) == expected
+            classDecl == expected
 
         where:
             decl             | expected

